evennia/contrib/aws-s3-cdn.py

```diff
@@ -0,0 +1,847 @@
+        clean_name = ''
+
+    return clean_name
+
+
+def safe_join(base, *paths):
+    """
+    A version of django.utils._os.safe_join for S3 paths.
+    Joins one or more path components to the base path component
+    intelligently. Returns a normalized version of the final path.
+    The final path must be located inside of the base path component
+    (otherwise a ValueError is raised).
+    Paths outside the base path indicate a possible security
+    sensitive operation.
+    """
+    base_path = force_text(base)
+    base_path = base_path.rstrip('/')
+    paths = [force_text(p) for p in paths]
+
+    final_path = base_path + '/'
+    for path in paths:
+        _final_path = posixpath.normpath(posixpath.join(final_path, path))
+        # posixpath.normpath() strips the trailing /. Add it back.
+        if path.endswith('/') or _final_path + '/' == final_path:
+            _final_path += '/'
+        final_path = _final_path
+    if final_path == base_path:
+        final_path += '/'
+
+    # Ensure final_path starts with base_path and that the next character after
+    # the base path is /.
+    base_path_len = len(base_path)
+    if (not final_path.startswith(base_path) or final_path[base_path_len] != '/'):
+        raise ValueError('the joined path is located outside of the base path'
+                         ' component')
+
+    return final_path.lstrip('/')
+
+
+def check_location(storage):
+    if storage.location.startswith('/'):
+        correct = storage.location.lstrip('/')
+        raise ImproperlyConfigured(
+            "{}.location cannot begin with a leading slash. Found '{}'. Use '{}' instead.".format(
+                storage.__class__.__name__,
+                storage.location,
+                correct,
+            )
+        )
+
+
+def lookup_env(names):
+    """
+    Look up for names in environment. Returns the first element
+    found.
+    """
+    for name in names:
+        value = os.environ.get(name)
+        if value:
+            return value
+
+
+def get_available_overwrite_name(name, max_length):
+    if max_length is None or len(name) <= max_length:
+        return name
+
+    # Adapted from Django
+    dir_name, file_name = os.path.split(name)
+    file_root, file_ext = os.path.splitext(file_name)
+    truncation = len(name) - max_length
+
+    file_root = file_root[:-truncation]
+    if not file_root:
+        raise SuspiciousFileOperation(
+            'aws-s3-cdn tried to truncate away entire filename "%s". '
+            'Please make sure that the corresponding file field '
+            'allows sufficient "max_length".' % name
+        )
+    return os.path.join(dir_name, "{}{}".format(file_root, file_ext))
+
+
+@deconstructible
+class S3Boto3StorageFile(File):
+
+    """
+    The default file object used by the S3Boto3Storage backend.
+    This file implements file streaming using boto's multipart
+    uploading functionality. The file can be opened in read or
+    write mode.
```

Comment:
This seems to be an internal thing used by the Storage backend. This should be clarified because the text makes it sound like it should be manually accessed (and if so, a usage example is needed - but since a file object is normally opened within a `with` context manager, there is no need to explicitly call `close` on it either).